### PR TITLE
[DeviceInfo] Caching values for device info.

### PR DESCRIPTION
**Pull request by Karthick-Somasundaresan**
Device info values are fetched via COMRPC for the first time and are stored in a cache further queries are fetched from the cache.

### CODE REVIEW

**Review by sramani-metro — APPROVED** [review]

File: Source/deviceinfo/DeviceInfo.cpp [hunk 1]
```diff
@@ -611,106 +614,574 @@ class DeviceInfoLink : public WPEFramework::RPC::SmartInterfaceType<WPEFramework
...
+
+    ~MemoryCachedDeviceInfo()
+    {
+    }
+
+public:
+    uint32_t Deviceinfo_architecure(char buffer[], uint8_t* length)
+    {
+        uint32_t result = Core::ERROR_NONE;
+        if(_architecture.IsSet())
```
Comment (MFransen69) [inline Source/deviceinfo/DeviceInfo.cpp:655]: As a coding guideline we always compare with something for readability. So in cases like this where you do not compare with something else specifically, we compare with == false or == true.
So in short this should be changed to if(_architecture.IsSet() == true).
Please also check for other occurrences in this file

File: Source/deviceinfo/DeviceInfo.cpp [hunk 2]
```diff
@@ -611,106 +614,574 @@ class DeviceInfoLink : public WPEFramework::RPC::SmartInterfaceType<WPEFramework
...
+        {
+            strncpy(buffer, _architecture.Value().c_str(), _architecture.Value().length());
+            *length = _architecture.Value().length();
```
Comment (MFransen69) [inline Source/deviceinfo/DeviceInfo.cpp:658]: See it is missing in most of the DeviceInfo code so I can imagine you never thought about it but it might be a good idea to add some asserts to check for pointers being passed not being null

File: Source/deviceinfo/DeviceInfo.cpp [hunk 3]
```diff
@@ -611,106 +614,574 @@ class DeviceInfoLink : public WPEFramework::RPC::SmartInterfaceType<WPEFramework
...
+        return result;
+
+    }
+
+    uint32_t Deviceinfo_id(uint8_t buffer[], uint8_t* length)
+    {
+        uint32_t result = Core::ERROR_NONE;
+        if(_id.size()>0)
+        {
+            std::copy(_id.begin(), _id.end(), buffer);
```
Comment (MFransen69) [inline Source/deviceinfo/DeviceInfo.cpp:717]: This could fail if *length was not big enough to hold the previously returned id array

File: Source/deviceinfo/DeviceInfo.cpp [hunk 4]
```diff
@@ -611,106 +614,574 @@ class DeviceInfoLink : public WPEFramework::RPC::SmartInterfaceType<WPEFramework
...
+            *length = _id.size();
+
+        }
+        else
+        {
+            result = _deviceInfoLink.Deviceinfo_id(buffer,length);
+            if(result == Core::ERROR_NONE)
+            {
+                _id.insert(_id.begin(), buffer, buffer + (*length));
```
Comment (MFransen69) [inline Source/deviceinfo/DeviceInfo.cpp:726]: you could optimize this a bit by before the copying doing a reserve on the std::vector to prevent multiple realocations, these tent to be costly

File: Source/deviceinfo/DeviceInfo.cpp [hunk 5]
```diff
@@ -611,106 +614,574 @@ class DeviceInfoLink : public WPEFramework::RPC::SmartInterfaceType<WPEFramework
...
+            if(result == Core::ERROR_NONE)
+            {
+                _maxOutputResolution = *value;
+            }
+        }
+        return result;
+    }
+
+    uint32_t Deviceinfo_output_resolutions(deviceinfo_output_resolution_t value[], uint8_t* length)
+    {
```
Comment (MFransen69) [inline Source/deviceinfo/DeviceInfo.cpp:946]: See comments at Deviceinfo_id, they most probably also reply here (and perhaps also at other functions below but will not repeat them)

File: Source/deviceinfo/DeviceInfo.cpp [hunk 6]
```diff
@@ -611,106 +614,574 @@ class DeviceInfoLink : public WPEFramework::RPC::SmartInterfaceType<WPEFramework
...
+        _hdr_atmos_cec |= 0x02;
+    }
+
+    void setCecSupport(const bool& cec)
+    {
+        _hdr_atmos_cec |= (1 & cec) << 5;
+        _hdr_atmos_cec |= 0x01;
+    }
+
+    bool getCecSupport()
```
Comment (MFransen69) [inline Source/deviceinfo/DeviceInfo.cpp:1024]: the getters can be const I guess

File: Source/deviceinfo/DeviceInfo.cpp [hunk 7]
```diff
@@ -611,106 +614,574 @@ class DeviceInfoLink : public WPEFramework::RPC::SmartInterfaceType<WPEFramework
...
+            {
+                _audioOutput.insert( _audioOutput.begin(), value, value + (*length));
+            }
+        }
+        return result;
+    }
+
+private:
+
+    void setHdrSupport(const bool& hdr)
```
Comment (MFransen69) [inline Source/deviceinfo/DeviceInfo.cpp:1006]: Why "const bool&" and not just "const bool"? (yes perhaps the bool in the second version is not what you are used to but we do that even for copy by value params to indicate that they are not supposed to change although that is would not happen anyway (and it saved us from some bugs when somebody by accident tried to change them and was not aware that would not propagate to the caller)

File: Source/deviceinfo/DeviceInfo.cpp [hunk 8]
```diff
@@ -611,106 +614,574 @@ class DeviceInfoLink : public WPEFramework::RPC::SmartInterfaceType<WPEFramework
...
+    {
+        _hdr_atmos_cec |= (1 & hdr) << 7;
```
Comment (MFransen69) [inline Source/deviceinfo/DeviceInfo.cpp:1008]: Not very important but what is the use of the "1 &"? Think it also works the same without it

File: Source/deviceinfo/deviceinfo.h [hunk 9]
```diff
@@ -24,7 +24,8 @@ typedef enum deviceinfo_hdcp_type {
     DEVICEINFO_HDCP_14,
     DEVICEINFO_HDCP_20,
     DEVICEINFO_HDCP_21,
-    DEVICEINFO_HDCP_22
+    DEVICEINFO_HDCP_22,
+    DEVICEINFO_HDCP_LENGTH
```
Comment (MFransen69) [inline Source/deviceinfo/deviceinfo.h:28]: Perhaps I miss something but why was this added?

**Review by MFransen69 — CHANGES_REQUESTED** [review]

**Review by MFransen69 — CHANGES_REQUESTED** [review]

File: Source/deviceinfo/DeviceInfo.cpp [hunk 10]
```diff
@@ -178,155 +197,301 @@ class DeviceInfoLink : public WPEFramework::RPC::SmartInterfaceType<WPEFramework
...
+                strncpy(buffer, modelName.c_str(), *length);
             }
-
+            *length = static_cast<uint8_t>(size + 1);
         } else {
-            *length=0;
+            *length = 0;
         }
-
+        _lock.Unlock();
```
[the remark this answers is not in the document]
Comment (Karthick-Somasundaresan) [reply Source/deviceinfo/DeviceInfo.cpp:311]: We are caching the value in the above else part, which I thought should be protected as well hence I took the lock ahead.

File: Source/deviceinfo/deviceinfo.h (hunk 9 above)
Comment (Karthick-Somasundaresan) [reply Source/deviceinfo/deviceinfo.h:28]: The other Enums had LENGTH as the last value, but only this one didn't have. Hence for uniformity's sake, I added it.

File: Source/deviceinfo/DeviceInfo.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (MFransen69) [reply Source/deviceinfo/DeviceInfo.cpp:311]: Ah sorry I was unclear, I meant moving the Unlock up, not the Lock down :) I think the unlock can be moved to above the  if (result == Core::ERROR_NONE ) but perhaps I'm wrong :)

File: Source/deviceinfo/deviceinfo.h (hunk 9 above)
Comment (MFransen69) [reply Source/deviceinfo/deviceinfo.h:28]: Good reason, I like consistency ;)

**Review by MFransen69 — CHANGES_REQUESTED** [review]

File: Source/deviceinfo/DeviceInfo.cpp [hunk 11]
```diff
@@ -380,37 +579,65 @@ class DeviceInfoLink : public WPEFramework::RPC::SmartInterfaceType<WPEFramework
...
+                uint8_t tempBuffer[255] = {'\0'};
+                uint8_t size = _identifierInterface->Identifier(sizeof (tempBuffer) - 1, tempBuffer);
+                std::copy(tempBuffer, tempBuffer + size, std::back_inserter(_id));
+                _id.shrink_to_fit();
+                result = Core::ERROR_NONE;
+            }
+        }
+        if (result == Core::ERROR_NONE)
+        {
+            *length = (_id.size() > (*length) - 1)? (*length) - 1: _id.size();
```
[the remark this answers is not in the document]
Comment (Karthick-Somasundaresan) [reply Source/deviceinfo/DeviceInfo.cpp:606]: Yes, I've re-verified that the behaviour is in line with the previous version.

**Review by MFransen69 — APPROVED** [review]
